modin/core/storage_formats/pandas/query_compiler.py

```diff
@@ -470,34 +477,53 @@ def map_func(left, right=right, kwargs=kwargs):
                     keep_partitioning=False,
                 )
             )
-            is_reset_index = True
-            if left_on and right_on:
-                left_on = left_on if is_list_like(left_on) else [left_on]
-                right_on = right_on if is_list_like(right_on) else [right_on]
-                is_reset_index = (
-                    False
-                    if any(o in new_self.index.names for o in left_on)
-                    and any(o in right.index.names for o in right_on)
-                    else True
-                )
-                if sort:
+
+            # Here we want to understand whether we're joining on a column or on an index level.
+            # It's cool if indexes are already materialized so we can easily check that, if not
+            # it's fine too, we can also decide that by columns, which tend to be already
+            # materialized quite often compared to the indexes.
+            keep_index = False
+            if self._modin_frame._index_cache is not None:
+                if left_on is not None and right_on is not None:
+                    keep_index = any(
+                        o in self.index.names
+                        and o in right_on
+                        and o in right.index.names
+                        for o in left_on
+                    )
+                elif on is not None:
+                    keep_index = any(
+                        o in self.index.names and o in right.index.names for o in on
```

Comment:
```suggestion
                        o in self.index.names and (o in right_on or o in right.index.names) for o in on
```
Should we check this way?

Reply:
no, see [the comment above](https://github.com/modin-project/modin/pull/5619/files#r1100073715)